### PR TITLE
RFC: PyTorch DevLog

### PR DESCRIPTION
**Pull request by laithsakka**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #178721
# Proposal: PyTorch DevLog

## TL;DR

Create a `devlog/` directory in the PyTorch repo where developers publish
technical notes as Markdown files via normal PRs.  Organized per
team/subsystem, with a deliberately lower bar than the official
pytorch.org/blog, and automatically accessible to AI coding assistants and
OSS contributors.

## Problem

PyTorch developers regularly write deep technical content — design
discussions, performance analyses, directional proposals, feature
introductions — and share them as ephemeral posts.  This content has a
practical lifespan of days to weeks before it is buried under the daily
volume of new posts.  The consequences:

- **Knowledge is lost.**  A post explaining FSDP2 memory management,
  TorchDynamo guard specialization, or export graph serialization becomes
  undiscoverable within a month.
- **OSS contributors never see it.**  Most of these discussions are purely
  technical with no reason to stay internal, yet they never reach the
  open-source community.
- **AI coding assistants can't use it.**  LLM tools that index the PyTorch
  repo (Copilot, etc.) have no way to learn from content that isn't
  checked in.

## Proposal

Add a `devlog/` directory to the PyTorch repository, open to any team:

```
pytorch/
└── devlog/
    ├── README.md
    ├── compile/
    │   ├── dynamo/
    │   ├── inductor/
    │   └── dynamic_shapes/
    ├── distributed/
    │   ├── fsdp/
    │   ├── dtensor/
    │   └── c10d/
    ├── export/
    ├── core/
    │   ├── autograd/
    │   └── dispatcher/
    └── ...
```

Posts are plain Markdown files following a simple naming convention
(`YYYY-MM-DD-short-title.md`).  Publishing is just a PR.  To start a new
area, create `devlog/<your-area>/README.md` and open a PR.

### 1. AI-accessible context
LLM coding assistants already index the PyTorch repo.  DevLog entries
become part of the context these tools draw on when answering questions
about PyTorch internals — no special integration needed.

#### AI Quality and Response Time: A Case Study

I asked: "How did PyTorch solve the data-dependent error problem?"

Without devlog — [P2256431852](https://www.internalfb.com/intern/paste/P2256431852) 9 minutes.139.5k tokens Mediocre, incomplete answer mentioning related tools/APIs without connecting them.

With devlog — [P2256394622](https://www.internalfb.com/intern/paste/P2256394622) 38 seconds. Correct answer, higher quality. 
Identified : "The fix: Explicit Unbacked Semantics"
Cited sources: "Per the 2025-10-29 Slaying Framework Data-Dependent Errors…" 
End "This answer is based on devlog entries. Would you like me to search the code to verify or supplement?"

### 2. OSS visibility
The PyTorch community extends well beyond Meta.  Technical write-ups about
FSDP internals, Dynamo tracing, export serialization, or autograd edge
cases are stuff external contributors need.  Hosting them in-repo
makes them discoverable via GitHub search, linkable from issues and PRs,
and citable in external discussions.

### 3. Durability and organization
Markdown files in the repo are versioned, searchable, and permanent.
Ephemeral posts fade within days.  Per-subsystem directories make it
trivial to answer "what's new with FSDP?" or "what's new with dynamic
shapes?" — just check the relevant README.

## Process

1. Create a Markdown file using the provided `_template.md`.
2. Place it in the appropriate topic directory
   (e.g., `devlog/compile/dynamo/`, `devlog/distributed/fsdp/`).
3. Open a PR — reviewers check technical accuracy and readability.
4. Merge.  Done.

No build system changes.  No website deployment.  No CI integration.  Just
Markdown files in a directory.

## Example posts (already converted)

The compile team has seeded five posts under `compile/dynamic_shapes/`:

| Date | Title | Author |
|------|-------|--------|
| 2026-03-25 | [Unbacked Dynamic Shapes Shouldn't Be Slower — Now They Aren't](./2026-03-25-unbacked-perf-parity.md) | @laithsakka |
| 2026-02-27 | [Reducing Compile-Time Overhead in Unbacked-Symbol-Heavy torch.export Traces](./2026-02-27-compile-time-unbacked-export.md) | @laithsakka, Aditya Venkataraman |
| 2026-01-20 | [Backed to Unbacked: From Guardable to Guardless Shapes](./2026-01-20-backed-to-unbacked.md) | @laithsakka, @aditvenk, @bobrenjc93 |
| 2025-10-29 | [Slaying the Framework Data-Dependent Errors Dragon](./2025-10-29-slaying-framework-ddes.md) | @laithsakka |
| 2025-07-08 | [Guard-Free Dynamic Shapes](./2025-07-08-guard-free-dynamic-shapes.md) | @laithsakka, @bdhirsh, @angelayi, @colinpeppler, @bobrenjc93, @avikchaudhuri, @aorenste, @pianpwk |
## Next steps

1. Land the initial directory structure and seed posts.
2. Announce and encourage contributions from all PyTorch teams.

### CODE REVIEW

File: devlog/compile/dynamic_shapes/2026-01-20-backed-to-unbacked.md [hunk 1]
```diff
@@ -0,0 +1,218 @@
+---
+title: "Backed to Unbacked: From Guardable to Guardless Shapes in PyTorch"
+author: Laith Sakka (@laithsakka), Aditya Venkataraman, Bob Ren
```
Comment (laithsakka) [inline devlog/compile/dynamic_shapes/2026-01-20-backed-to-unbacked.md]: add github accounts of other authors.

File: devlog/compile/dynamic_shapes/2026-01-20-backed-to-unbacked.md [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+date: 2026-01-20
+tags: [dynamic_shapes, unbacked, backed, torch.compile, frontier]
+---
+
+# Backed to Unbacked: From Guardable to Guardless Shapes in PyTorch
+
+![Hint stamp](./images/2026-01-20-backed-unbacked-header.jpg)
+
+> **TL;DR**
```
Comment (laithsakka) [inline devlog/compile/dynamic_shapes/2026-01-20-backed-to-unbacked.md]: why two  **TL;DR**?

File: devlog/compile/dynamic_shapes/2026-01-20-backed-to-unbacked.md [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+> **TL;DR** – We expect unbacked dynamic shapes to become the dominant
+> shape mechanism for Frontier-style workloads due to their better
+> predictability and controllability.  However, some blockers remain —
+> most notably the performance gap, which is a primary focus for the first
+> half of 2026.
+
```
Comment (laithsakka) [inline devlog/compile/dynamic_shapes/2026-01-20-backed-to-unbacked.md:33]: add in this post ... bla bla bla what is this post abput?

File: devlog/compile/dynamic_shapes/2026-03-25-unbacked-perf-parity.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+# Unbacked Dynamic Shapes Shouldn't Be Slower — Now They Aren't
+
+![Backed vs unbacked perf parity](./images/2026-03-25-perf-parity-header.jpg)
+
+> **TL;DR** – Unbacked dynamic shapes had 2x–20% slowdowns on TorchBench
+> and ~30% regressions on vLLM.  We fixed the root causes — now unbacked
+> matches backed across all tested models and configurations.
+
+## Motivation
```
Comment (laithsakka) [inline devlog/compile/dynamic_shapes/2026-03-25-unbacked-perf-parity.md:16]: this is wrong !

**Review by jansel — CHANGES_REQUESTED** [review]
I like the idea of docs, but wonder if it should go in pytorch/pytorch or a different github repo?

The other option would be to use https://github.com/pytorch/rfcs -- which seems like a very similar idea.

My worry would be this would suffer the same fate as pytorch/rfcs and be mostly abandoned and not used.  If it is going to be abandoned I would prefer that happen in a different repo.

Doing it in pytorch/pytorch would also incur more CI cost and bloat the repo size.

**Comment by ezyang** [comment]
obsoleted by https://github.com/pytorch/devlogs
